### PR TITLE
[BugFix] Fix `pad_sequence` behavior for non-tensor attributes of tensorclass

### LINKED ISSUE #783: [BUG] `tensordict.pad_sequence` silently ignores non-tensor attributes in `tensorclass`es or `TensorDict`s
## Describe the bug

I have some `tensorclass`es that store an audio file, with some metadata including speaker id and utterance id. I would like to collate these `tensorclass`es to form a batch, however when I do so, the metadata is lost (the metadata from the first tensordict is kept for every item in the batch) and the user is not warned about this either.

## To Reproduce

Steps to reproduce the behavior.

```python
from tensordict import pad_sequence, TensorDict

d1 = TensorDict({'a': torch.tensor([0]), 'b': ['asd']})
d2 = TensorDict({'a': torch.tensor([0]), 'b': ['efg']})

pad_sequence([d1, d2])
```

```bash
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([2, 1]), device=cpu, dtype=torch.int64, is_shared=False),
        b: NonTensorData(data=asd, batch_size=torch.Size([2]), device=None)},
    batch_size=torch.Size([2]),
    device=None,
    is_shared=False)
```

## Expected behavior

I should either get a properly joined tensordict, e.g., 
```
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([2, 1]), device=cpu, dtype=torch.int64, is_shared=False),
        b: NonTensorData(data=['asd', 'efg'], batch_size=torch.Size([2]), device=None)}, # in a list same shape as the batch_size
    batch_size=torch.Size([2]),
    device=None,
    is_shared=False)
```
or `tensordict.pad_sequence` should warn the user that the metadata is being discarded.

## System info

```
tensordict-nightly              2024.5.18
```

## Reason and Possible fixes

If you know or suspect the reason for this bug, paste the code lines and suggest modifications.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/tensordict/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

### PR DESCRIPTION
**Pull request by kurtamohler**
## Description

Fix the way `pad_sequence` checks for non-tensor attributes of tensorclasses. Also raise an error if `pad_sequence` is given tensorclasses with `return_mask=True`, since we cannot add the "masks" attribute with `<tensorclass>.set()`.

## Motivation and Context

close #783

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/tensordict/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tensordict/functional.py [hunk 1]
```diff
@@ -149,6 +156,13 @@ def pad_sequence(
     if not list_of_tensordicts:
         raise RuntimeError("list_of_tensordicts cannot be empty")
 
+    if return_mask and is_tensorclass(list_of_tensordicts[0]):
+        raise RuntimeError(
```
Comment (vmoens) [inline tensordict/functional.py:160]: Yep that's also my take on this! 👍

File: tensordict/functional.py [hunk 2]
```diff
@@ -163,7 +177,7 @@ def pad_sequence(
     for i, td in enumerate(list_of_tensordicts):
 
         for key in keys:
-            item = td.get(key)
+            item = td._tensordict.get(key) if is_tensorclass(td) else td.get(key)
```
Comment (vmoens) [inline tensordict/functional.py:180]: is_tensorclass is a bit too expensive, let's take this check out of the loop (isolate the tensordict we'll be using before iterating through the keys)

**Review by vmoens — COMMENTED** [review]
Just one little thing to fix but otherwise LGTM!

File: tensordict/functional.py [hunk 3]
```diff
@@ -161,9 +175,10 @@ def pad_sequence(
     list_of_dicts = [{} for _ in range(len(list_of_tensordicts))]
     keys_copy = list(keys)
     for i, td in enumerate(list_of_tensordicts):
+        is_tensorclass_ = is_tensorclass(td)
```
Comment (vmoens) [inline tensordict/functional.py:178]:
```suggestion
        if is_tensorclass(td):
            td = td._tensordict
```

File: tensordict/functional.py [hunk 4]
```diff
@@ -161,9 +175,10 @@ def pad_sequence(
...
 
         for key in keys:
-            item = td.get(key)
+            item = td._tensordict.get(key) if is_tensorclass_ else td.get(key)
```
Comment (vmoens) [inline tensordict/functional.py:181]:
```suggestion
            item = td.get(key)
```

**Review by vmoens — APPROVED** [review]
Thanks for this!
